src/handlers/contact_handlers.py

```diff
@@ -1,140 +1,190 @@
-    
+
     else:
         record = Record(name)
         record.add_phone(phone)
         book.add_record(record)
-    
+
     return "Contact added."
 
 
 @input_error
 def add_phone(args: list[str], book: AddressBook) -> str:
-    pass
+    name, phone = args
+    record: Record = book.find(name, True)
+
+    if phone in map(lambda phone: phone.value, record.phones):
+        return "Phone already added."
+
+    record.add_phone(phone)
+
+    return "Phone added."
 
 
 @input_error
 def add_birthday(args: list[str], book: AddressBook) -> str:
     name, birthday = args
-    record: Record = book.find(name)
+    record: Record = book.find(name, True)
 
-    if record is None:
-        raise KeyError(name)
-    
     record.add_birthday(birthday)
-    
-    return "Birthday added."
+
+    return "Birthday updated."
 
 
 @input_error
 def add_email(args: list[str], book: AddressBook) -> str:
-    pass
+    name, email = args
+    record: Record = book.find(name, True)
+
+    record.add_email(email)
+
+    return "Email updated."
 
 
 @input_error
 def add_address(args: list[str], book: AddressBook) -> str:
-    pass
+    name, *address = args
+    record: Record = book.find(name, True)
+
+    record.add_address(" ".join(address))
+
+    return "Address updated."
 
 
 @input_error
 def change_phone(args: list[str], book: AddressBook) -> str:
     name, old_phone, new_phone = args
-    record: Record = book.find(name)
+    record: Record = book.find(name, True)
 
-    if record is None:
-        raise KeyError(name)
-    
     record.edit_phone(old_phone, new_phone)
-    
-    return "Contact updated."
+
+    return "Phone updated."
 
 
 @input_error
-def change_birthday(args: list[str], book: AddressBook) -> str:
-    pass
+def delete_contact(args: list[str], book: AddressBook) -> str:
+    name = args[0]
+    book.delete(name)
+
+    return "Contact removed."
 
 
 @input_error
-def change_email(args: list[str], book: AddressBook) -> str:
-    pass
+def delete_phone(args: list[str], book: AddressBook) -> str:
+    name, phone = args
+    record: Record = book.find(name, True)
 
+    if len(record.phones) < 2:
+        return "Unable to delete last phone number from contact."
 
-@input_error
-def change_address(args: list[str], book: AddressBook) -> str:
-    pass
+    record.remove_phone(phone)
 
+    return "Phone removed."
 
-@input_error
-def delete_contact(args: list[str], book: AddressBook) -> str:
-    pass
 
+@show_all_error
+def show_all(args: list[str], book: AddressBook) -> str:
+    if not book:
+        raise NotFoundError
+
+    if bool(len(args)):
+        search_fragment = args[0]
+        found_records = [
+            record
+            for record in sorted(book.values(), key=lambda record: record.name.value)
+            if search_fragment in record.email.value
+        ]
 
-@input_error
-def delete_phone(args: list[str], book: AddressBook) -> str:
-    pass
+        if not bool(len(found_records)):
```

Comment:
```suggestion
        if not len(found_records):
```
І так до булевого типу приводится після if

Reply:
done